### DISCUSSION TITLE
How to automate application using Browser's API window.showDirectorypicker with cypress and recognize the selected directory as a valid one

### DISCUSSION THREAD

**Discussion by PradeepaParandhaman** [cypress-io/cypress#29711] [Questions and Help]
In our application Input field is blocked due to security reasons and I've tried mock directory selection and also some python/AutoIt scripts to create a directory and utilize it in the tests. Still it was not recognized as an actual directory. 

Please let me know if this issue is addressed anywhere.

**Comment by portyu9** [comment]
Hi @PradeepaParandhaman 

Creating a directory on disk is not enough to make it equivalent to a directory selected through:

```js
window.showDirectoryPicker()
```

`showDirectoryPicker()` is part of the browser's **File System Access API**. It displays browser/native UI and returns a `FileSystemDirectoryHandle` carrying browser-managed permission/state.

That is why an externally created folder from Python/AutoIt is not automatically recognized by the page as if `showDirectoryPicker()` had returned it.

There are two different testing goals here.

### 1. Test your application's behavior after a directory was chosen

For this, abstract or stub `showDirectoryPicker()` and return the minimum fake handle your application needs.

For example:

```js
cy.visit('/', {
  onBeforeLoad(win) {
    cy.stub(win, 'showDirectoryPicker').resolves({
      kind: 'directory',
      name: 'test-directory',

      async *values() {
        // return fake directory entries if your application uses them
      },
    })
  },
})
```

The exact fake object needs to implement whichever `FileSystemDirectoryHandle` methods your application actually calls.

An even cleaner design is to wrap the browser API in your own function/module and inject or stub that dependency in tests.

### 2. Test the real browser/native directory chooser end-to-end

Cypress does not provide a normal command for selecting a directory from the native UI created by `window.showDirectoryPicker()`.

The browser intentionally protects this API: it requires a secure context and **transient user activation**, and the returned handle is browser-controlled.

`cy.selectFile()` is useful for:

```html
<input type="file">
```

and drag/drop file workflows, but it is not a replacement for `window.showDirectoryPicker()`.

So for a Cypress test I would normally:

1. keep one thin layer around `showDirectoryPicker()`;
2. stub that layer/API in Cypress;
3. test all application behavior using a realistic fake `FileSystemDirectoryHandle`;
4. separately cover the browser's own picker integration manually or with tooling specifically capable of automating that native/browser API.

Trying to create the folder externally and then force Cypress to make the browser believe the user chose it bypasses browser permission state and is therefore not equivalent.

Please mark this answer as accepted if it helped, thank you.
